internal/locald/sandboxmanager/sandbox_manager.go

```diff
@@ -194,16 +202,42 @@ func (m *sandboxManager) Run(ctx context.Context) error {
 		}
 	}
 
+	// Start devbox session manager
+	m.devboxSessionMgr.Start(runCtx)
+
 	// Run the sandboxes watcher
 	sbmWatcher.run(runCtx, m.tunAPIClient)
 
 	// Wait until termination
 	<-runCtx.Done()
 
-	// Clean up
+	// Check if shutdown was triggered by devbox session release
+	if m.devboxSessionMgr != nil && m.devboxSessionMgr.WasSessionReleased() {
+		m.log.Info("Devbox session was released, entering deadend state")
+		m.sessionReleased = true
+		
+		// Shutdown root manager (tunnel, localnet, etchosts)
+		m.shutdownRootManager()
+		
+		// Stop all active work but keep gRPC server running
+		sbmWatcher.stop()
+		m.devboxSessionMgr.Stop(ctx)
+		if m.portForward != nil {
+			m.portForward.Close()
+		} else if m.ctlPlaneProxy != nil {
+			m.ctlPlaneProxy.Close(ctx)
+		}
```

Comment:
We could normalize this into a func (it's dup a few lines below).

Reply:
I would agree if it encapsulated some self-contained thing, but it's a partial shutdown (not including grpc server), and it seems odd to have a separate func to me.

Happy to do it if you insist, but would prefer not b/c of the above.